typegraph/core/src/runtimes/prisma/type_generation/with_filters.rs

```diff
@@ -0,0 +1,231 @@
+                        });
+                    }
+                    Type::String(_) => {
+                        props.push(Prop {
+                            key: k.to_string(),
+                            type_id: id,
+                            prop_type: PropType::String,
+                        });
+                    }
+                    _ => {
+                        return Err(format!(
+                            "type '{}' not supported by prisma",
+                            ty.get_data().variant_name()
+                        ));
+                    }
+                }
+            }
+            Ok(())
+        })?;
+
+        let mut builder = t::struct_();
+        builder.named(self.name(context));
+        for prop in props {
+            let ty = match prop.prop_type {
+                PropType::Boolean => context.generate(&CompleteFilter(BooleanFilter))?,
+                PropType::Number(num) => context.generate(&CompleteFilter(NumberFilter(num)))?,
+                PropType::String => context.generate(&CompleteFilter(StringFilter))?,
+                PropType::Object => context.generate(&WithFilters {
+                    type_id: prop.type_id,
+                    skip_rel: true,
+                })?,
+            };
+            builder.prop(prop.key, t::optional(ty).build()?);
+        }
+
+        Ok(builder.build()?)
+    }
+
+    fn name(&self, _context: &TypeGenContext) -> String {
+        let suffix = if self.skip_rel { "_norel" } else { "" };
+        format!("_{}WithFilters{}", self.type_id.0, suffix)
+    }
+}
+
+struct CompleteFilter<T: TypeGen>(T);
+
+impl<T: TypeGen> TypeGen for CompleteFilter<T> {
+    fn generate(&self, context: &mut TypeGenContext) -> Result<TypeId> {
+        let inner = context.generate(&self.0)?;
+        t::optional(t::union([inner, t::struct_().prop("not", inner).build()?]).build()?)
+            .named(self.name(context))
+            .build()
+    }
+
+    fn name(&self, context: &TypeGenContext) -> String {
+        format!("{}_c", self.0.name(context))
+    }
+}
+
+struct BooleanFilter;
+
+impl TypeGen for BooleanFilter {
+    fn generate(&self, context: &mut TypeGenContext) -> Result<TypeId> {
+        t::union([
+            t::boolean().build()?,
+            t::struct_().prop("equals", t::boolean().build()?).build()?,
+        ])
+        .named(self.name(context))
+        .build()
+    }
+
+    fn name(&self, context: &TypeGenContext) -> String {
+        "_boolean_filter".to_string()
+    }
+}
+
+enum NumberType {
+    Integer,
+    Float,
+}
+
+struct NumberFilter(NumberType);
+
+impl TypeGen for NumberFilter {
+    fn generate(&self, context: &mut TypeGenContext) -> Result<TypeId> {
+        let type_id = match self.0 {
+            NumberType::Integer => t::integer().build()?,
+            NumberType::Float => t::float().build()?,
+        };
+        let opt_type_id = t::optional(type_id).build()?;
+        let array_type_id = t::array(type_id).build()?;
```

Comment:
Are these store id or the canonical index?

Reply:
Store IDs.
At this stage, we only have the store ids.